### ISSUE TITLE
Cannot find happy executable in build plan

### ISSUE THREAD

**Issue by snoyberg** [commercialhaskell/stack#3733]
Actions: create a new project (e.g., with `stack new foo`), add a `build-tool` dependency on `happy`, and run `stack build`.

Expected: installs happy and then our package

Actual: Stack complains that:

```
No packages found in snapshot which provide a "happy" executable, which is a build-tool dependency of "foo"
```

**Comment by TheWizardTower** [comment]
Soooo... I can't reproduce this. :(

Steps:

cd ~/
stack new foo
cd foo
vim package.yaml (add "build-tools: happy" to the 'executables:' section)
stack build

Result: Happy is not built in the package. Running `find . -type f -executable` in ~/foo only shows foo executables, inside ~/foo/.stack-work .

No warnings or messages are printed about this, however, so this is a bug -- just not the one @snoyberg is describing above.

I observed this behavior on stack 1.6.1 and 1.6.3. @snoyberg helped me through this process, and wasn't able to get me to the point where I was reproducing the bug.

**Comment by TheWizardTower** [comment]
Oh, now _this_ is exciting.

When I ran the above steps, I created the foo project using stack 1.6.1, snapshot lts-10.2.

I just rm'd the dir and re-ran the steps, now that I've upgraded to 1.6.3.

Happy was built.


` m@VirtualFamiliar  ~/foo  stack build
mtl-2.2.1: configure              
mtl-2.2.1: build                  
Cabal-2.0.1.1: configure          
Cabal-2.0.1.1: build              
mtl-2.2.1: copy/register          
Cabal-2.0.1.1: copy/register      
happy-1.19.8: configure           
happy-1.19.8: build               
happy-1.19.8: copy/register       
Building all executables for `foo' once. After a successful build of all of them, only specified executables will be rebuilt.           
foo-0.1.0.0: configure (lib + exe)
Configuring foo-0.1.0.0...        
foo-0.1.0.0: build (lib + exe)    
Preprocessing library for foo-0.1.0.0..                             
Building library for foo-0.1.0.0..
[1 of 2] Compiling Lib              ( src/Lib.hs, .stack-work/dist/x86_64-linux-tinfo6/Cabal-2.0.1.0/build/Lib.o )                      
[2 of 2] Compiling Paths_foo        ( .stack-work/dist/x86_64-linux-tinfo6/Cabal-2.0.1.0/build/autogen/Paths_foo.hs, .stack-work/dist/x86_64-linux-tinfo6/Cabal-2.0.1.0/build/Paths_foo.o )                 
Preprocessing executable 'foo-exe' for foo-0.1.0.0..                
Building executable 'foo-exe' for foo-0.1.0.0..                     
[1 of 2] Compiling Main             ( app/Main.hs, .stack-work/dist/x86_64-linux-tinfo6/Cabal-2.0.1.0/build/foo-exe/foo-exe-tmp/Main.o )
[2 of 2] Compiling Paths_foo        ( .stack-work/dist/x86_64-linux-tinfo6/Cabal-2.0.1.0/build/foo-exe/autogen/Paths_foo.hs, .stack-work/dist/x86_64-linux-tinfo6/Cabal-2.0.1.0/build/foo-exe/foo-exe-tmp/Paths_foo.o )                                                         
Linking .stack-work/dist/x86_64-linux-tinfo6/Cabal-2.0.1.0/build/foo-exe/foo-exe ...                                                    
foo-0.1.0.0: copy/register        
Installing library in /home/merlin/foo/.stack-work/install/x86_64-linux-tinfo6/lts-10.3/8.2.2/lib/x86_64-linux-ghc-8.2.2/foo-0.1.0.0-Cx6ymLwVUGUJkxkFB5qTEW               
Installing executable foo-exe in /home/merlin/foo/.stack-work/install/x86_64-linux-tinfo6/lts-10.3/8.2.2/bin                            
Registering library for foo-0.1.0.0..                               
Completed 4 action(s).            
`

Note that the happy executable still isn't in the project dir:

` m@VirtualFamiliar  ~/foo  find . -executable -type f
./.stack-work/install/x86_64-linux-tinfo6/lts-10.3/8.2.2/bin/foo-exe
./.stack-work/install/x86_64-linux-tinfo6/lts-10.3/8.2.2/lib/x86_64-linux-ghc-8.2.2/libHSfoo-0.1.0.0-Cx6ymLwVUGUJkxkFB5qTEW-ghc8.2.2.so 
./.stack-work/dist/x86_64-linux-tinfo6/Cabal-2.0.1.0/build/foo-exe/foo-exe                                                              
./.stack-work/dist/x86_64-linux-tinfo6/Cabal-2.0.1.0/build/libHSfoo-0.1.0.0-Cx6ymLwVUGUJkxkFB5qTEW-ghc8.2.2.so  `

**Comment by mgsloan** [comment]
I also cannot reproduce this as specified.  However, I did run into this when doing `stack build hlint` from within the stack source repo.  The issue is that stack specifies an extra-dep for Cabal, which promotes something like half the snapshot to local.

I tried a straightforward fix to the problem, but found that it would require a more substantial change, which I've described in https://github.com/commercialhaskell/stack/issues/3777 .  Closing this in favor of that issue.
